extensions/s3/src/main/java/io/deephaven/extensions/s3/S3RequestCache.java

```diff
@@ -57,36 +57,37 @@ int getFragmentSize() {
      * @return the request
      */
     @NotNull
-    S3ChannelContext.Request.AcquiredRequest getOrCreateRequest(@NotNull final S3Uri uri, final long fragmentIndex,
+    Request.AcquiredRequest getOrCreateRequest(@NotNull final S3Uri uri, final long fragmentIndex,
             @NotNull final S3ChannelContext context) {
         final Request.ID key = new Request.ID(uri, fragmentIndex);
         Request.AcquiredRequest newAcquiredRequest = null;
-        Request existingRequest = requests.get(key);
-        for (int retryCount = 0; retryCount < Integer.MAX_VALUE; retryCount++) {
+        while (true) {
+            final Request existingRequest = requests.get(key);
             if (existingRequest != null) {
                 final Request.AcquiredRequest acquired = existingRequest.tryAcquire();
                 if (acquired != null) {
                     return acquired;
-                } else {
-                    remove(existingRequest);
                 }
             }
             if (newAcquiredRequest == null) {
                 newAcquiredRequest = Request.createAndAcquire(fragmentIndex, context);
             }
-            existingRequest = requests.putIfAbsent(key, newAcquiredRequest.request);
+            final boolean added;
             if (existingRequest == null) {
+                // Ideally, we could have used ".replace" in this case as well, but KeyedObjectHashMap.replace currently
+                // has a bug when the key is not present in the map.
+                added = requests.putIfAbsent(key, newAcquiredRequest.request) == null;
```

Comment:
When this fails, we can avoid `get`.

Reply:
You found this more complicated. I still prefer to avoid an extra lookup. Maybe moot anyway, since we can't seem to trust `replace` right now.
```
        Request existingRequest = requests.get(key);
        while (true) {
            if (existingRequest != null) {
                final Request.AcquiredRequest acquired = existingRequest.tryAcquire();
                if (acquired != null) {
                    return acquired;
                }
            }
            if (newAcquiredRequest == null) {
                newAcquiredRequest = Request.createAndAcquire(fragmentIndex, context);
            }
            final boolean added;
            if (existingRequest == null) {
                added = (existingRequest = requests.putIfAbsent(key, newAcquiredRequest.request)) == null;
            } else {
                if (!(added = requests.replace(key, existingRequest, newAcquiredRequest.request))) {
                    existingRequest = requests.get(key);
                }
            }
```

Reply:
Reverted back to `remove` + `putIfAbsent` pattern and added a TODO for this with issue #5486.